### DISCUSSION TITLE
Support of MAX7219 for LED dot matrix

### DISCUSSION THREAD

**Discussion by sternegugger** [arendst/Tasmota#13590] [Ideas]
Hi developers,

I have a 32x8 LED dot matrix based on MAX7219.
It can be extended with more 8x8 dot modules in rows and columns.
Reading the tasmato documentation I found the module MAX7219, but it only supports 7 segment displys.
(define USE_DISPLAY_MAX7219)
When I connect a 8x8 dot matrix to MAX7219 CLK. DN and CS and send "DisplayText 1" it only shows two dots in the first line, which represents the two segments for the number 1.
The documentation contains the setting: DisplayModel 3 (= 8x8 Dot-LED-Matrix).
But using this, it is ignored and I get allways "RESULT = {"DisplayModel":15}"

I also found the option USE_DISPLAY_MATRIX, which needs an HT16K33 Dot-Matrix.
Searching on Amazon and other platforms I mostly found LED dot matrixes based on MAX7219.

Now my question:

- Is it possible to combine the tasmota dot matrix functionality with the MAX7219 modul?
- If it is allready implemented, which compile options do I have to use?

Data sheets:
 - [MAX7219 data sheet](https://www.openimpulse.com/blog/wp-content/uploads/wpsc/downloadables/MAX7219-Datasheet.pdf)
 - [MAX7219 32x8 Dot Matrix Display Module](https://www.handsontec.com/dataspecs/module/MAX7219-4-In-1.pdf)

## Solved:
See documentation: [MAX7219 for 8x8 LED dot matrix](https://tasmota.github.io/docs/MAX7219/)

**Comment by Rushmed** [comment]
I am very interested too!

**Comment by sternegugger** [comment]
Hi Rushmed,

in the meanwhile I developed the MAX272 8x8 dot matrix driver by myself.
See Pull Request [#14091](https://github.com/arendst/Tasmota/pull/14091)
![20211204_133243](https://user-images.githubusercontent.com/24937182/144709719-2fb4ff2c-7645-4dd8-b0a3-eccd306d63f0.jpg)
![20211204_133418](https://user-images.githubusercontent.com/24937182/144709723-833ff002-ef84-431a-b05f-8b7ef30e0072.jpg)

The driver enables the display of ascii text on MAX7219 based 8x8 LED dot matrix modules (1088AS).

**#define USE_DISPLAY_MAX7219_MATRIX**
and build tasmota_display

  Connect the MAX7219 display module's pins to any free GPIOs of the ESP8266 or ESP32 module.
  VCC should be 5V. Depending on the number of used modules and the brightness, the used current can be more than 500 mA.
  Connect the 5 outgoing pins (VCC, GND, DI, CS, CLK) of the first module to the next one. 

  With this you can connect up to 32 modules. 
  To extend the display height, multiple rows are supported. Each module row starts from left to right.
  Assign the pins as follows from Tasmota's GUI:

  DIN hardware pin --> "MAX7219 DIN"
  CS hardware pin --> "MAX7219 CS"
  CLK hardware pin --> "MAX7219 CLK"

  Once the GPIO configuration is saved and the ESP8266/ESP32 module restarts,
  set the Display Model to 19 and Display Mode to 0

  Depending on the order of the wired 8x8 matrix modules you have got a display of size pixel_width x pixel_height.
  The size has to be set with the commands "DisplayWidth <pixel_width>" and "DisplayHeight <pixel_height>"

  After the ESP8266/ESP32 module restarts again, turn ON the display with the command "Power 1"
  Now, the following "Display" commands can be used:

  **DisplayText**  text 
    Sends the text to the display. 
    If the text fits into the display, it is shown in the center.
    Otherwise it scrolls to the left and repeats as long it is cleared or new "DisplayText" overwrites it.

  **DisplayDimmer** [0..100]
    Sets the intensity of the display.

  **DisplayBlinkrate** [0..3]
    0: not blinking
    1: slow, 2: medium 3: fast blinking

  **Power** [ON|OFF]
    Sitches the display on or off. When "off", the display buffer is not cleared and will be shown again when after "Power ON". 
    Other display commands are still active when off.

  **DisplayClear**
    Clears the display

  **DisplayScrollDelay** [0..15]   // default = 0
    Sets the speed of text scroll. Smaller delay = faster scrolling.
    The maximum scroll speed is 50ms per pixel on DisplayScrollDelay 0.

  **DisplayWidth** [8..256]
    Sets the pixel width of the display (8x number of modules in a row)

  **DisplayHeight** [8..256]
    Sets the pixel height of the display (8x number of module rows)

  **DisplayRotate** [0|2]
    0: normal orientation; devide 0 starts at top left
    2: upside down; device 0 starts at bottom right

  **DisplayClock**  [0|1|2]
    Displays a clock.
    Commands "DisplayClock 1"     // 12 hr format
            "DisplayClock 2"     // 24 hr format
            "DisplayClock 0"     // turn off clock; please use additional cammand: DisplayMode 0

  If you would like to use the UTF8 latin1 character set, it cam be added by copile option: 
  #define USE_UTF8_LATIN1

A feedback after you tried it would be great!

Yours Michael
(sternegugger)

**Reply by nimannv** [reply]
Hi @sternegugger.

First of all, thank you for your excellent explanation.
I did it, and everything worked well, though the display crashed when I disconnected the board power and connect it again, it appears that something gets wrong with the display led addressing after powering off and on the board.

Does anybody have any advice?
My board: ESP Wemos D1

**Reply by sternegugger** [reply]
@nimannv 
When you power the display OFF and ON again, it has not been initialized by the ESP. In this case you have to restart the ESP after the display is on again.

**Reply by nimannv** [reply]
@sternegugger 
I'm not sure what kind of restart you mean, however, I did "Restart 1" and it seems nothing happened after that, and I still see this:

![sorry](https://user-images.githubusercontent.com/12507587/201542786-4f1d64cf-1032-4ad4-a677-8fe81543824b.jpeg)

The only way that I've found to fix this is to configure everything again by "reset 1" (reset device settings to firmware defaults and restart)

**Reply by sternegugger** [reply]
@nimannv 
Restart method 1: power off and on the ESP
Restart method 2: Press the restart button on the ESP board (if exists)
Restart method 3: Click "Restart" button in the main menu at the web frontend
![image](https://user-images.githubusercontent.com/24937182/201712340-3ee6ef99-b7f3-49b7-9b5a-3729e1d77e1f.png)
Restart method 4: Send the MQTT command "restart 1" 

The Display will be initalized after ESP restart. The display must be ON to be initialized.

**Reply by montyfert** [reply]
hi if anyone could help it would be great ,
i have the MAX7219 conected to d5 d7 d8 
installed the display firmware 
![image](https://user-images.githubusercontent.com/74438819/211339617-60c94a3a-f46d-4193-bd60-28595f725d59.png)
it wont let me change DisplayModel to 19 
![image](https://user-images.githubusercontent.com/74438819/211340003-9297c254-982e-4bdd-9d62-b53202b64288.png)
im new to tasmota and dont have a clue lol.
thanks 
martin

**Reply by rug1024** [reply]
You **must** see these lines beginning with MTX in the console at boot.
![grafik](https://user-images.githubusercontent.com/96115731/211343426-fedec1d4-1c9b-4aaf-97c3-e72769099291.png)
If not, you have probably not compiled a firmware including the MAX7219 drivers.

**Reply by montyfert** [reply]
"compiled a firmware including the MAX7219 drivers."
i would'nt know how to lol
can i not USE the tasmota display bin

**Reply by rug1024** [reply]
No, MAX7219 is not included. It's yet a bit special.
Tell me what hardware you have, I'll compile one for you. (Wemos, ESP8266, ESP32, ... ?)

**Reply by montyfert** [reply]
thank you 
its a d1 mini

**Reply by rug1024** [reply]
Here you go.... Tasmota Dev (Wemos 4K) with the here shown options:

![grafik](https://user-images.githubusercontent.com/96115731/211398411-0a2b8094-8045-408d-bb0f-a07813609c68.png)

![grafik](https://user-images.githubusercontent.com/96115731/211397938-a2889bfd-6f93-445f-ad26-82a6719daa80.png)

Done with https://gitpod.io/#https://github.com/benzino77/tasmocompiler/tree/development

Download Firmware: https://upload.disroot.org/r/q4a2HuLy#FXqGCC77HG2yUdNxElpl1Hxg6H5sZeAdJa1b55jhm0c=

Happy testing!

**Reply by montyfert** [reply]
thanks but i must be doing something wrong 
![image](https://user-images.githubusercontent.com/74438819/211407369-728141cb-52b3-446a-bfb6-3bf32283fcd5.png)

**Reply by barbudor** [reply]
Yes. You must first upgrade to Tasmota-minimal.bin.gz to free-up some flash space. Then upgrade with the provided bin.gz

**Reply by montyfert** [reply]
ive done it i installed tasmota lite the upgraded

**Reply by rug1024** [reply]
No problem.
Flash 'tasmota-minimal' first, then flash my file.

**Reply by montyfert** [reply]
thanks for the help
martin

**Reply by montyfert** [reply]
ive just got to figure out what to send via mqtt to display something lol

**Reply by sfromis** [reply]
First play around using commands in the console. Then such commands can be directly used via MQTT too.

**Reply by montyfert** [reply]
thanks

**Reply by sternegugger** [reply]
> hi if anyone could help it would be great , i have the MAX7219 conected to d5 d7 d8 installed the display firmware it wont let me change DisplayModel to 19  im new to tasmota and dont have a clue lol. thanks martin

Hi Martin,

to use display model 19, you have to compile tasmota code with this option:
`#define USE_DISPLAY_MAX7219_MATRIX`
(modify file: `user_config_override.h`)
and build tasmota_display manually.

see also: https://tasmota.github.io/docs/Compile-your-build/#general-customization-principle

**Reply by montyfert** [reply]
me again lol 
uptill yesterday everything working great had weather icon in the middle mqtt temps either side .
today had to set it up again .

**Reply by Martin12350** [reply]
Hi Guys, pretty awesome PR, but one small detail that would maybe solve a problem for others (or possibly future me).
DisplayRotate allows values from **0 to 3**, not only _0 or 2_. And in my case the display driver have probably differently wired columns, which result in each display 8x8 showing the text in reverse. Setting the value to 3 all works as expected! Which is awesome, but gave me a hard time to figure it out, since in the doc only 2 possible values are mentioned. After searching how other libraries doing it and digging in the source code, finally noticed, it is simple as that. So updating doc with small hint, that if display is acting "weirdly", it is possible that the driver is routed differently. Here is a pretty good article that put me on the right track: [HW wiring issue](https://arduinoplusplus.wordpress.com/2017/04/14/parola-a-to-z-adapting-for-different-hardware/)

**Reply by sfromis** [reply]
The docs pages has a small pencil icon top right in the main column, if you want to suggest improvements :smiley:

**Reply by Martin12350** [reply]
Do not know if the missing 1/3 rotation options have some meaning :) 
E.g. not finished feature or will totally break something and should not be used yet :D

**Reply by sfromis** [reply]
Not involved in the driver, I "know" nothing, can only observe that the underlying library has 4 orientation values defined.
https://github.com/arendst/Tasmota/blob/5cbe5fe9e4999133e81ebd026a6f775382381ad7/lib/lib_display/LedControl/src/LedMatrix.h#L58-L61

**Comment by pcarlos** [comment]
Thanks.
I's a perfect solution.

**Comment by rug1024** [comment]
I intend to use 2 Matrix-Displays on one ESP8266.
Display1 should always show the clock and Display2 should show another value.
Is this possible?

**Reply by sternegugger** [reply]
Currently only one display line is supported.
Sending text to a second line is not implemented yet.

Possible workaround: Use on line. The first half for the time string the second half for the other value.
E.g. you have 2x 4 Matrix Modules. Connect them as a single line and use it as two displays.
`DisplayWidth 32`
`DisplayHeight 8`
`DisplayText 12:59  other`
You have to build the time string with your own application.
This workaround is a bit tricky. You have to fill spaces in the middle to have a fix string length.

**Comment by ozett** [comment]
hi @sternegugger ,
playing with berry-scripting on ESP32 with tasmota 12.x i was wondering if your driver works with berry-scripting-language?
is it possible to move dots with berry-scripting?

---edit:
mayby with lvgl and uDisplay? on a one-dot-level (x,y)? that would be cool..

**Comment by gishanc** [comment]
Could you explain a bit more how it works when 2 modules connected as 2 lines (32 width and 16 height in pixels for example)
Since it supports only 1 line, does it make the font larger to make use of both modules?

**Comment by 4920441** [comment]
Great work, thanks a lot!
Are there any plans to adapt smaller fonts as well?

**Reply by sfromis** [reply]
"Plans" is not how things are done, but instead modifications "open source style". As things are now, the code contains one font without a mechanism for selecting among multiple.

**Reply by 4920441** [reply]
yes, sure, but if someone is 99% through with the plans and it is nearly implemented, it does not make too much sense to repeat the work again, or does it?

It probably does, but ist does not have to be :-)

**Comment by ScooterOD** [comment]
Hallo,

I cannot change displaymodel to 19.

RSL: INFO1 = {"Info1":{"Module":"max7219","Version":"12.5.0(display)","FallbackTopic":"cmnd/DVES_106B62_fb/","GroupTopic":"cmnd/tasmotas/"}}

CMD: display
RSL: RESULT = {"Display":{"Model":15,"Type":1,"Width":32,"Height":1,"Mode":0,"Dimmer":20,"Size":1,"Font":1,"Rotate":0,"Invert":0,"Refresh":1,"Cols":[32,8],"Rows":1}}

CMD: displaymodel 19
RSL: RESULT = {"DisplayModel":15}

D2 GPIO4 = MAX7219 DIN
D1 GPIO5 = MAX7219 CS
D5 GPIO14 = MAY7219 CLK

Display is a DOTMatrix, not 7-segment, so I need to change it

MAX7219 is listed at both Models. Is this right? How will it differentiated? 

I use a Wimo-D1. Now with Tasmota-Display Version 12.5.
I changed  to that little bit older Version to test.
With 13.0.1 release and lastest it's the same behavior.

Do you have any hints?

Many thanks

CU
Michael

**Reply by sfromis** [reply]
For dot matrix MAX7219, the docs page https://tasmota.github.io/docs/MAX7219/ clearly states that 
> To use this driver, tasmota has to be compiled with:
> `#define USE_DISPLAY_MAX7219_MATRIX`

Until that, there is no DisplayModel 19 available.

**Reply by ScooterOD** [reply]
Yes I found (and understand ;-) ) it now.
I will compile it for myself. II thought I could save that for myself.
Many thanks

CU
Michael

**Reply by viktor6** [reply]
What firmware should be for ESP 32?

**Reply by sfromis** [reply]
As I mentioned, you need to do your own custom build of Tasmota for this device support. Not hard to do.

**Reply by viktor6** [reply]
I don't know how to create my own build for esp32

**Reply by viktor6** [reply]
I'm trying to create my own
https://gitpod.io#https://github.com/arendst/Tasmota/tree/master
user_config_override.h
before
#endif // _USER_CONFIG_OVERRIDE_H_
I add
#define USE_DISPLAY_MAX7219_MATRIX
I save
I'm launching
platformio run -e tasmota32

What am I doing wrong?

![Opera Снимок_2023-11-21_112858_arendst-tasmota-5i8iyxgx92x ws-eu106 gitpod io](https://github.com/arendst/Tasmota/assets/10570918/fa7edfe8-4d20-4ff7-8f9b-2b2daee71da6)

**Reply by viktor6** [reply]
Should I do this or not?

![Opera Снимок_2023-11-21_115912_3000-benzino77-tasmocompiler-6nyri08q5ot ws-eu106 gitpod io](https://github.com/arendst/Tasmota/assets/10570918/73434b32-644f-44b7-b6d3-80ac25a371a5)

**Reply by rug1024** [reply]
When using TasmoCompiler that would be the right way.

**Comment by drtorchwood** [comment]
If I see it correctly, the upper seven dots are used for the text and the bottom row is always dark. Is it possible to turn on the lowest line as well? 
Background: I would like to display two different numbers on the display and the use the bottom row as an indicator which one is currently displayed.

**Reply by sfromis** [reply]
Some lower-case letters, comma, semicolon and underscore do extend into the bottom row.
